### PR TITLE
Add Checks Specification Documentation

### PR DESCRIPTION
**Pull request by nelsonkopliku**
# Description

This PR resumes a bit what was started here https://github.com/trento-project/web/pull/674.

Some documentation has been added relative to the Checks Specification

go here for a more visual experience :point_right: https://github.com/trento-project/wanda/blob/checks_documentation/guides/specification.md

### CODE REVIEW

**Comment by abravosuse** [comment]
@nelsonkopliku I was reviewing https://github.com/trento-project/wanda/blob/checks_documentation/guides/dsl.md. So far so good. But went out, came back and the link doesn't work now:
image
What happened?

File: guides/specification.md [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+- [Values](#values)
+- [Expectations](#expectations)
+- [Expression Language](#expression-language)
+    - [Evaluation Scope](#evaluation-scope)
+
+## Introduction
+
+The need this DSL aims to fulfill is to provide users a simple way to declare what we (the Trento Core Team) often refer to as "Checks". 
+
+Checks are, in Trento's domain, the crystallization of SUSE's best practices when it comes to SAP workloads in a form that both a man and a machine can read.
```
Comment (arbulu89) [inline guides/specification.md:28]: I wouldn't use the `man` termine. Maybe better `human`

File: guides/specification.md [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+    - [Facts, Values, Expectations](#facts-values-expectations)
+- [Facts](#facts)
+- [Values](#values)
+- [Expectations](#expectations)
+- [Expression Language](#expression-language)
+    - [Evaluation Scope](#evaluation-scope)
+
+## Introduction
+
+The need this DSL aims to fulfill is to provide users a simple way to declare what we (the Trento Core Team) often refer to as "Checks".
```
Comment (arbulu89) [inline guides/specification.md:26]: Do we need to reference the `Trento Core Team` here?
I would refer to `Trento Team` in any case

File: guides/specification.md [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+- [Expression Language](#expression-language)
+    - [Evaluation Scope](#evaluation-scope)
+
+## Introduction
+
+The need this DSL aims to fulfill is to provide users a simple way to declare what we (the Trento Core Team) often refer to as "Checks". 
+
+Checks are, in Trento's domain, the crystallization of SUSE's best practices when it comes to SAP workloads in a form that both a man and a machine can read.
+
+We get several benefits from this approach:
```
Comment (arbulu89) [inline guides/specification.md:30]: Usually, using the `we` term is not recommended in the docs (at least, our suse docs team says so). 
I would maybe change to `This approach provides several benefits`

File: guides/specification.md [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+
+A Check declaration comes in the form of a `yaml` file and all the Checks together build up the **Checks Catalog**
+
+Here's an example:
+```yaml
+id: 156F64
+name: Corosync `token` timeout
+group: Corosync
+description: |
+  Corosync `token` timeout is set to `{{ values.expected_token_timeout }}`
```
Comment (arbulu89) [inline guides/specification.md:46]: I would remove this renderization in the description. 
We don't have it (and I honestly thing we shouldn't have it).
Let's play safe

File: guides/specification.md [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+ie: 
+```yaml
+name: Corosync `token` timeout
+name: Corosync `consensus` timeout
+name: SBD Startmode
+```
+
+**NOTES:** 
+- in the current ansible implementation the name is set to something like `1.1.1`, `1.1.2`, ...
+- in the current wanda implementation we are not really using this
```
Comment (arbulu89) [inline guides/specification.md:119]: It is used in the `Checks selection` view at least

File: guides/specification.md [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+  The value of the Corosync `token` timeout is not set as recommended.
+  ## Remediation
+  Adjust the corosync `token` timeout as recommended on the best 
+  ...
+  2. Reload the corosync configuration:
+  ...
+
+```
+
+#### Severity
```
Comment (arbulu89) [inline guides/specification.md:194]: In small letter to be consistent

File: guides/specification.md [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+```yaml
+facts:
+  - name: corosync_token_timeout
+    gatherer: corosync.conf
+    argument: totem.token
+
+  # other facts maybe
+```
+
+Finally, gathered facts, will be used in Check's [Expectations](#expectations) to determine whether expected conditions are met for the best practice to be adhered.
```
Comment (arbulu89) [inline guides/specification.md:256]: Let's try to not use future term. Something like `Finally, gathered facts, are used in...`

File: guides/specification.md [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+
+- there may be many conditions
+- first condition that passes determines the value, following are not evaluated
+- `when` entry [Expression](#expression-language) has [access](#evaluation-scope) to gathered [facts](#facts-1) and [env](#env) evaluation scopes
+
+All the _resolved_ declared values would be registered in the [`values`](#values-1) namespaced evaluation scope.
+
+## Expectations
+
+Expectations section is where the assumptions about the state of a target infra for a given scenario are formalized in expressions determining whether a check passes or not.
```
Comment (arbulu89) [inline guides/specification.md:368]: `infra` to `infrastructure` better

File: guides/specification.md [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+> ```
+
+Considering the previous scenario what happens is that:
+- the fact `corosync_token_timeout` is gathered on all targets (`A` and `B` in this case)
+- the expectation expression gets executed against the `corosync_token_timeout` fact gathered on every targets.
+  - `targetA.corosync_token_timeout == values.expected_token_timeout`
+  - `targetB.corosync_token_timeout == values.expected_token_timeout`
+- every evaluation has to be `true`
+
+### Expect Same
```
Comment (arbulu89) [inline guides/specification.md:438]: The same, maybe `expect_same` better

File: guides/specification.md [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+- the fact `awesome_fact` is gathered on all targets (`A` and `B` in this case)
+- the expectation expression gets executed for every target involved.
+  - `targetA.facts.awesome_fact`
+  - `targetB.facts.awesome_fact`
+- the expressions results has to be the same for every target
+  - `targetA.facts.awesome_fact == targetB.facts.awesome_fact`
+
+## Expression Language
+
+Different parts of the Check declaration are places where a computation is needed.
```
Comment (arbulu89) [inline guides/specification.md:480]: Should we use `evaluation` instead of `computation`?

File: guides/specification.md [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+
+> Provide a contextual [Description](#description)/[Remediation](#remediation) text, based on the execution context
+> 
+> interpolation of `{{ <expression> }}`
+```yaml
+description: |
```
Comment (arbulu89) [inline guides/specification.md:486]: Sorry for being insistent, but I wouldn't put interpolations in descriptions as example

File: guides/specification.md [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+
+Every expression has access to an evaluation scope, allowing to access relevant piece of information to run the expression.
+
+Scopes are namespaced and access to items in the scope is name based.
+
+#### **env**
+
+`env` is a map of information about the context of the running execution, it is set by the system on each execution/check compilation.
+
+Available entries in scope
```
Comment (arbulu89) [inline guides/specification.md:527]: I would put this maybe as an example. Wanda doesn't know about which are the coming env values. Actually, it accepts everything coming in the execution request and it depends on the domain logic

File: guides/gatherers.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+## Available Gatherers
+
+Here's a collection of build-in gatherers, with information about how to use them.
+
+| name                          | implementation          
+| ----------------------------- | ------------------------------------------
+| [`corosync.conf`](#corosync)  | [trento-project/agent/../gatherers/corosyncconf.go](https://github.com/trento-project/agent/blob/main/internal/factsengine/gatherers/corosyncconf.go)
+
+### Corosync
```
Comment (arbulu89) [inline guides/gatherers.md:27]: Let's use `corosync.conf` here. We have other gatherers about corosync (`corosynccmapctl` which returns runtime corosync configuration values)

File: guides/gatherers.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+**Name:** `corosync.conf`
+
+This gatherer allows accessing information contained in `/etc/corosync/corosync.conf`
+
+```
+corosync_related_fact = corosync.conf(some.argument)
+```
+
+Accepted arguments
```
Comment (arbulu89) [inline guides/gatherers.md:37]: Those looks like examples to me. They are more options (way more)

File: guides/expression_language.md [hunk 15]
```diff
@@ -0,0 +1,167 @@
+# Trento Checks Expression Language Reference
+
+A small, fast, easy-to-use scripting language and evaluation engine
```
Comment (arbulu89) [inline guides/expression_language.md:3]: Should we put some reference to the rhai book?
It might help the users to find advanced use cases
https://rhai.rs/book/language/

**Review by arbulu89 — COMMENTED** [review]
@nelsonkopliku Fantastic work!
Really advanced documentation.
I have added many comments, most of them are about removing not implemented features from the docs. Others simple suggestions.

I would like to wait until some other end-user can give a look. I'm still biased with a lot of information, so I might've missed many things.

File: guides/specification.md (hunk 1 above)
Comment (nelsonkopliku) [reply guides/specification.md:28]: thanks :alien: streamlined to `users` if you're ok.

File: guides/gatherers.md (hunk 14 above)
Comment (nelsonkopliku) [reply guides/gatherers.md:37]: Changing to `Sample arguments`, anyway there are links to the implementation and tests.
If we want to add more, we can iterate over.

File: guides/expression_language.md (hunk 15 above)
Comment (nelsonkopliku) [reply guides/expression_language.md:3]: Added here adb85e15eb5d24e6269b37950fc799c779fc125a

File: guides/expression_language.md [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+### Comparing different types defaults to `false`
+
+Comparing two values of _different_ data types defaults to `false`.
+
+The exception is `!=` (not equals) which defaults to `true`. This is in line with intuition.
+
+```rust
+42 > "42";          // false: i64 cannot be compared with string
+42 <= "42";         // false: i64 cannot be compared with string
+let ts = new_ts();  // custom type
```
Comment (fabriziosestito) [inline guides/expression_language.md:77]: what is a custom type? are the checks implementers interested in this?

File: guides/expression_language.md [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+- [Variables](#variables)
+- [Logic Operators](#logic-operators)
+- [If Statement](#logic-operators)
+- [Rhai]
+## Introduction
+
+[NameToBeDefined] is an embedded scripting language and evaluation engine for Trento Checks Expressions that gives a safe and easy way to script specific steps during Checks Execution.
+
+## Types
+TBD
```
Comment (fabriziosestito) [inline guides/expression_language.md:18]: Let's try to address all the TBD parts prior merging

File: guides/specification.md [hunk 18]
```diff
@@ -0,0 +1,549 @@
...
+    argument: totem.token
+
+  # other facts maybe
+```
+
+Finally, gathered facts, are used in Check's [Expectations](#expectations) to determine whether expected conditions are met for the best practice to be adhered.
+
+## Values
+
+Values are basically named variables that may evaluate differently based on the execution context and are used with Facts for _Contextual_ [Expectations](#expectations) Evaluation.
```
Comment (fabriziosestito) [inline guides/specification.md:238]: remove "basically"

File: guides/specification.md [hunk 19]
```diff
@@ -0,0 +1,549 @@
...
+
+The need this DSL aims to fulfill is to provide users a simple way to declare what we (the Trento Team) often refer to as "Checks". 
+
+Checks are, in Trento's domain, the crystallization of SUSE's best practices when it comes to SAP workloads in a form that both a user and a machine can read.
+
+This approach provides several benefits:
+
+- Users can formalize best practices with no space for ambiguity;
+- Machines can assert systems' state, automatically, with no space for ambiguity;
+- The development of new best practices gets streamlined through a common definition that allows to firestart shared efforts.
```
Comment (fabriziosestito) [inline guides/specification.md:34]: This intro does well in an RFC/ADR/design document or marketing document.
I think it's too fuzzy for documentation where we want to be more specific on what the application does, I would rephrase it in simpler terms without stating the benefits.
The focus should be on how it works rather than why we did it.

Also, there is no high-level explanation of how the check flow works.
E.g. execution requested -> facts gathered -> expectation evaluated.
I think this is really important for whoever implements a check to understand the checks lifetime.

File: guides/expression_language.md [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+
+```rust
+42 == 42.0;         // true
+42.0 == 42;         // true
+42.0 > 42;          // false
+42 >= 42.0;         // true
+42.0 < 42;          // false
+```
+
+### Strings interoperate with characters
```
Comment (fabriziosestito) [inline guides/expression_language.md:57]: I think this part is a bit too much complicated, and it also doesn't mention operations that are common when building a check such as arrays or maps operations.

File: guides/specification.md [hunk 21]
```diff
@@ -0,0 +1,549 @@
...
+> - it has to be `dog` when the `color` in the execution context is `blue`
+> - it has to be `rabbit` in all other cases, regardless of the execution context
+> 
+> so we define a named variable `awesome_expectation` that resolves to `cat|dog|rabbit` when proper conditions are met
+>
+> allowing us to have an expectation like this 
+>
+> `expect: facts.awesome_fact == values.expected_token_timeout`
+
+In a nutshell, Values provide a way to decouple concepts, usually expeced values for facts, from their actual concretization during execution in certain scenarios.
```
Comment (fabriziosestito) [inline guides/specification.md:287]: I find this sentense too complicated. Values are just variables.

File: guides/specification.md [hunk 22]
```diff
@@ -0,0 +1,549 @@
...
+
+- there may be many conditions
+- first condition that passes determines the value, following are not evaluated
+- `when` entry [Expression](#expression-language) has [access](#evaluation-scope) to gathered [facts](#facts-1) and [env](#env) evaluation scopes
+
+All the _resolved_ declared values would be registered in the [`values`](#values-1) namespaced evaluation scope.
+
+## Expectations
+
+Expectations section is where the assumptions about the state of a target infrastructure for a given scenario are formalized in expressions determining whether a check passes or not.
```
Comment (fabriziosestito) [inline guides/specification.md:346]: Could we rephrase this in something like
"Expectations are assertions on the state of a target infrastructure. By using fact and values they are able to determine if a check passes or not."

**Review by fabriziosestito — CHANGES_REQUESTED** [review]
Some comments on the specification part, which to me could be good enough to iterate on. I feel there is a lot of work to do still on the expression language yet.

File: guides/specification.md (hunk 21 above)
Comment (nelsonkopliku) [reply guides/specification.md:287]: Gonna remove the sentence, it's redundant indeed.

File: guides/specification.md (hunk 19 above)
Comment (nelsonkopliku) [reply guides/specification.md]: Changed here 4af3b8dab590c427e4938491c842a4d300d9b1fa

File: guides/check_definition.schema.json [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+          "default"
+        ]
+      },
+      "Condition": {
+        "title": "Condition",
+        "type": "object",
+        "additionalProperties": false,
+        "properties": {
+          "value": {
+            "type": "integer"
```
Comment (arbulu89) [inline guides/check_definition.schema.json:107]: The `value` could be a string as well

File: guides/check_definition.schema.json [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+          },
+          "gatherer": {
+            "type": "string"
+          },
+          "argument": {
+            "type": "string"
+          }
+        },
+        "required": [
+          "argument",
```
Comment (arbulu89) [inline guides/check_definition.schema.json:66]: I'm not sure if the `argument` would be always required, but well, let it be by now. It is fine

File: guides/specification.md [hunk 25]
```diff
@@ -0,0 +1,588 @@
...
+#### Facts, Values, Expectations
+
+See main sections [Facts](#facts), [Values](#values), [Expectations](#expectations)
+
+## Facts
+
+Facts are various types of information that the engine can discover on the target infrastructure.
+Examples include (but are not limited to) installed packages, cluster state, and configuration files' content.
+
+The process of determining the value of a declared fact during Check execution is refered to as _Facts Gathering_ and it is the job of the [_Gatherers_](./gatherers.md).
```
Comment (arbulu89) [inline guides/specification.md:242]: `refered` -> `referred`

File: guides/specification.md [hunk 26]
```diff
@@ -0,0 +1,588 @@
...
+> - use [**hardcoded**](#hardcoded-values) values
+> - define `values` as [**constants**](#constant-values)
+> 
+> Scenario:
+> 
+> No matter what the context is, the fact `awesome_fact` MUST always be `wanda`
+
+### Hardcoded Values
+
+Direct usage of a simple hadrcoded value
```
Comment (arbulu89) [inline guides/specification.md:291]: typo `hadrcoded`

File: guides/expression_language.md [hunk 27]
```diff
@@ -166,6 +115,158 @@ let x = if decision { 42 }; // no else branch defaults to '()'
...
+};
+
+some_map.foo // 42
+some_map.bar // "hello"
+
+```
+
+### Non-existing property
+
+Trying to read a non-existing property returns `()` instead of causing an error.
```
Comment (fabriziosestito) [inline guides/expression_language.md:228]: this actually raises in our expression language

**Review by arbulu89 — APPROVED** [review]
Hey @nelsonkopliku ,
Some additional comments added for the json schema.
I'm approving the PR without really having a look on the expression language, as I think you are looking it with @fabriziosestito ,
Besides that, I find the main specification and the gatherers good.

**Review by fabriziosestito — CHANGES_REQUESTED** [review]

File: guides/check_definition.schema.json [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+        ]
+      },
+      "Expectation": {
+        "title": "Expectation",
+        "type": "object",
+        "additionalProperties": false,
+        "minProperties": 2,
+        "maxProperties": 2,
+        "properties": {
+          "name": {
```
[the remark this answers is not in the document]
Comment (nelsonkopliku) [reply guides/check_definition.schema.json:125]: The representation you are referring to is the internal one we use in wanda :thinking: 

In this context each expectation has only a name and either an `expect` or `expect_same`

```yaml
expectations:
  - name: token_timeout
    expect: facts.corosync_token_timeout == values.expected_token_timeout
```

Works?

File: guides/check_definition.schema.json [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+            "type": "string"
+          }
+        },
+        "patternProperties": {
+          "^expect$|^expect_same$": {
```
[the remark this answers is not in the document]
Comment (nelsonkopliku) [reply guides/check_definition.schema.json:130]: I am afraid that works only for values, not for properties.

What we are saying is that next to `name` we expect either an `expect` or `expect_same`.

File: guides/check_definition.schema.json (hunk 23 above)
Comment (nelsonkopliku) [reply guides/check_definition.schema.json:107]: good catch!

File: guides/check_definition.schema.json (hunk 28 above)
[the remark this answers is not in the document]
Comment (arbulu89) [reply guides/check_definition.schema.json:132]: ok, forget what I said XD

File: guides/expression_language.md (hunk 27 above)
Comment (nelsonkopliku) [reply guides/expression_language.md]: Ah thanks :heart: ! Changed!

**Review by fabriziosestito — APPROVED** [review]
